File: backtracking/SubsetsWithDup90.java
import java.util.*;
public class SubsetsWithDup90{
    public static void main(String[] args) {
        int[] nums={4,4,4,1,4};
        SubsetsWithDup90 s=new SubsetsWithDup90();
        System.out.println(s.subsetsWithDup(nums));
    }

    public List<List<Integer>> subsetsWithDup(int[] nums) {
        if (nums==null || nums.length<=0) {
            return res;
        }
        Arrays.sort(nums);
        subsets(nums,0,new ArrayList());
        return res;
    }

    private  List<List<Integer>> res=new ArrayList<>();
    
    public void subsets(int[] nums,int index,List<Integer> lis) {
        res.add(new ArrayList(lis));
        for (int i=index;i<nums.length;i++) {
            if (i>index && nums[i] == nums[i-1]) {
                continue;
            }
            lis.add(nums[i]);
            subsets(nums,i+1,lis);
            lis.remove(lis.size()-1);
        }
    }

    //4 4 4 1 4
    //1 4 4 4 4
    public void subsets2(int[] nums,int index,List<Integer> lis) {
        HashSet<Integer> set=new HashSet<>();
        res.add(new ArrayList(lis));
        for (int i=index;i<nums.length;i++) {
            if (set.contains(nums[i])) {
                continue;
            }
            lis.add(nums[i]);
            set.add(nums[i]);
            subsets2(nums,i+1,lis);
            lis.remove(lis.size()-1);
        }
    }
}

File: backtracking/SubsetsWithDup90.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        nums = sorted(nums)
        res = []
        for mask in range(1<<n):
            tmp = []
            flag = 1
            for i in range(n):
                if (mask>>i)&1:
                    # 当前位被选取，前一位未被选取，且前一位和当前位相同
                    # 两者选一个的时候只有一种情况，10和01是同一种
                    # 或者 if i < n-1 and not (mask>>(i+1))&1 and nums[i] == nums[i+1]:
                    if i > 0 and not (mask>>(i-1))&1 and nums[i] == nums[i-1]:
                        flag = 0
                        break
                    tmp.append(nums[i])
            if flag:
                res.append(tmp)
        return res